text/0037-syft-bom.md

```diff
@@ -0,0 +1,103 @@
+# Support for CycloneDX and Syft SBoMs in Paketo
+
+## Summary
+
+Buildpacks which directly provide a dependency or buildpacks which install
+application dependencies will support the generation of SBoM in the format of
+CycloneDX and Syft. Per [CNB
+RFC#95](https://github.com/buildpacks/rfcs/blob/main/text/0095-sbom.md), the
+SBoM documents will live at `<layer>.bom.<ext>.json`, `launch.bom.<ext>.json`
+and `build.bom.<ext>.json` where `<ext>` will be `cdx.json` (CycloneDX) or
+`syft.json` (Syft).
+
+## Motivation
+
+[Paketo
+RFC#33](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0033-bill-of-materials.md)
+describes the initial strategy used to add support for SBoM generation in
+Paketo buildpacks. This RFC outlines the process by which Paketo buildpacks
+will build upon that initial strategy by including support for CycloneDX and
+Syft SBoM formats in accordance with the structure proposed in [CNB
+RFC#95:SBOM](https://github.com/buildpacks/rfcs/blob/main/text/0095-sbom.md).
+
+## Detailed Explanation
+
+Paketo buildpacks will now optionally produce SBoM for runtime &
+application dependencies in CycloneDX and Syft formats. Buildpack
```

Comment:
There are three different locations where a buildpack can put SBoM information. I don't feel like buildpacks RFC #95, provides any direction as to what should go in each of these files. 

Do we want to specify where the buildpacks should put certain types of information? To provide some consistency across buildpacks.

For example, SBoM information should be broken down like this:

- `<layer>.bom.<ext>.json` This file contains SBoM information specific to a layer. A Buildpack that installs dependencies will install them into a specific layer and should include the SBoM entry for the installed dependency into this layer-specific SBoM file.

- `launch.bom.<ext>.json` This file should contain SBoM information for dependencies not installed into a specific layer, such as application-specific dependencies that have been installed, and those dependencies will exist in the launch image.

- `build.bom.<ext>.json` This file should contain SBoM information for dependencies not installed into a specific layer, such as application-specific dependencies that have been installed, and those dependencies are only used at build time (i.e. they do not end up in the launch image).

I'm not sure if my example above is correct. I'm interested to see how others interpreted the RFC and what plans are for utilizing the different files. 

Regardless of how it's interpreted, it does seem worthwhile to me to provide some guidance, not requirements, on how these files should be used to promote consistency across buildpacks.

Reply:
That seems reasonable to me, I'll update this section

Reply:
This makes sense to me but I wonder whether we can be less strict in the guidance for `launch.bom` and `build.bom`, specifically the bit about the SBoM information being "for dependencies not installed to a layer". In the `node-module-bom` buildpack, for example, CycloneDX [is installed into a specific layer](https://github.com/paketo-buildpacks/node-module-bom/blob/cd06b3d2bee74891acc4135c0fa07c883f177103/build.go#L62) at build time, but its BoM is then appended to the general Build BoM as opposed to a layer-specific BoM.

As these are guidelines, would it be reasonable to you to omit that stipulation and state simply that launch.bom is for deps which end up in the app image and  build.bom is for deps contributed to the build environment?

Reply:
My initial comment was really just my initial interpretation of read through the spec and RFC #95. I'm absolutely not asserting that is correct. It's a hazy topic for me.

I'm totally fine with only using `launch.bom` and `build.bom`. It seems like everyone is on board with that and I think that distinction is clear and pretty easy. The spec also clearly defines this (not updated for RFC#95 but the spirit is the same):  

>SHOULD write BOM (Bill-of-Materials) entries to <layers>/launch.toml describing any contributions to the app image.
>SHOULD write build BOM entries to <layers>/build.toml describing any contributions to the build environment.

Where I get kind of confused is with the third file: `<layer>.bom.<ext>.json`. Given that build things go in `build.bom` and launch things go in `launch.bom`, what's left to go in `<layer>.bom`? Maybe nothing is the answer?

Reply:
OK. I asked for some clarifications in the buildpacks slack. There is some designed overlap in these files. The key benefits to using`<layer>.bom.<ext>.json` are this:

1. The SBoM information in this file is specifically attached to a particular layer, so it's more specific. It's not just saying, this software is in the generated image. It's saying, it's in the generated image and it was put into a specific layer by a specific buildpack, which created the layer.
2. If you put metadata into `<layer>.bom.<ext>.json` then it can be cached. This could potentially reduce the overhead of generating the SBoM metadata.

Given that, it sounds like we should prefer using `<layer>.bom.<ext>.json` for dependencies installed by buildpacks as that would allow the SBoM metadata to be cached with the layer. I think we should prefer or suggest, but not require that. That provides guidance but ultimately leaves it up to the buildpack author.

That would leave `build.bom` and `launch.bom` for the things that don't fit cleanly into layers, like possibly application dependencies or cases where it's not easy to differentiate between tooling and dependencies.

Does that seem reasonable?

Reply:
Just some background, rfc 95 superceded rfc 87 which introduced this concept. For more information see this section https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md#what-it-is

Reply:
@dmikusa-pivotal That does seem reasonable to me, I'll modify this section.

Reply:
@dmikusa-pivotal Updated the section with additional guidance